### ISSUE TITLE
Zed uses lot of RAM

### ISSUE THREAD

**Issue by tanaybhomia** [zed-industries/zed#18673]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Here zed is using around 2GB of ram and I have checked it on Reddit some people were saying their device is showing just 200mb to 300mb of ram usage with these kinds of project. 

### Environment

OS - Fedora 40 Workstation
Hardware Model - ASUSTeK COMPUTER INC. VivoBook
Processor - AMD Ryzen™ 7 5800HS 
Memory - 16GB
Gnome - 46 
Linux Kernel - Linux 6.10.11-200.fc40.x86_64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

![Screenshot from 2024-10-02 18-23-03](https://github.com/user-attachments/assets/de3e95f3-7969-43ea-bffe-fde873d0b297)
![Screenshot from 2024-10-02 18-24-09](https://github.com/user-attachments/assets/b6374ab8-fdb1-446f-92ce-ca07a0975c91)

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

2024-10-02T17:16:13.182467034+05:30 [INFO] ========== starting zed ==========
2024-10-02T17:16:13.332487855+05:30 [INFO] perform;
2024-10-02T17:16:13.332602972+05:30 [INFO] authenticate;
2024-10-02T17:16:13.332737114+05:30 [INFO] read_command;
2024-10-02T17:16:13.333205995+05:30 [INFO] new;
2024-10-02T17:16:13.333310091+05:30 [INFO] socket reader;
2024-10-02T17:16:13.333438412+05:30 [INFO] Opening main db
2024-10-02T17:16:13.333845798+05:30 [INFO] keep_updated;
2024-10-02T17:16:13.334505408+05:30 [INFO] Opening main db
2024-10-02T17:16:13.338805931+05:30 [INFO] Using git binary path: None
2024-10-02T17:16:13.364648978+05:30 [ERROR] theme not found: Adwaita Pastel Dark
2024-10-02T17:16:13.364837091+05:30 [ERROR] No such file or directory (os error 2) about ["/home/tanay/.config/github-copilot"]
2024-10-02T17:16:13.365633488+05:30 [INFO] initializing context server client
2024-10-02T17:16:13.365775405+05:30 [INFO] Registering built-in prompt template: content_prompt
2024-10-02T17:16:13.366111517+05:30 [INFO] Registering built-in prompt template: edit_workflow
2024-10-02T17:16:13.366391834+05:30 [INFO] Registering built-in prompt template: step_resolution
2024-10-02T17:16:13.366824808+05:30 [INFO] Registering built-in prompt template: terminal_assistant_prompt
2024-10-02T17:16:13.367069498+05:30 [INFO] No prompt template overrides directory found at /home/tanay/.local/share/zed/prompt_overrides. Using built-in prompts.
2024-10-02T17:16:13.367447498+05:30 [INFO] extensions updated. loading 2, reloading 0, unloading 0
2024-10-02T17:16:13.369345546+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.369452918+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.369539611+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.369625011+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.369784972+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.370766667+05:30 [INFO] activate is not implemented on Linux, ignoring the call
2024-10-02T17:16:13.37190141+05:30 [INFO] perform;
2024-10-02T17:16:13.372019051+05:30 [INFO] authenticate;
2024-10-02T17:16:13.372134468+05:30 [INFO] read_command;
2024-10-02T17:16:13.372243213+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.372371604+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.372482823+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.372670366+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.372787426+05:30 [INFO] socket reader;
2024-10-02T17:16:13.377726239+05:30 [INFO] Opening main db
2024-10-02T17:16:13.377847617+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.379215418+05:30 [INFO] new;
2024-10-02T17:16:13.379323902+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.380027575+05:30 [INFO] keep_updated;
2024-10-02T17:16:13.446684128+05:30 [INFO] Enabling Vulkan Portability
2024-10-02T17:16:13.446789796+05:30 [INFO] Enabling color space support
2024-10-02T17:16:13.464347479+05:30 [INFO] Adapter: "AMD Radeon Graphics (RADV RENOIR)"
2024-10-02T17:16:13.464490488+05:30 [INFO] No ray tracing extensions are supported
2024-10-02T17:16:13.479418468+05:30 [INFO] Using surface present mode MAILBOX
2024-10-02T17:16:13.4798559+05:30 [WARN] Unable to forbid exclusive full screen
2024-10-02T17:16:13.480583718+05:30 [INFO] Creating a descriptor pool for at most 16 sets
2024-10-02T17:16:13.480694626+05:30 [INFO] Creating a descriptor pool for at most 16 sets
2024-10-02T17:16:13.480795326+05:30 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: PreMultiplied }
2024-10-02T17:16:13.486594526+05:30 [INFO] Using GPU: GPUSpecs { is_software_emulated: false, device_name: "AMD Radeon Graphics (RADV RENOIR)", driver_name: "radv", driver_info: "Mesa 24.1.7" }
2024-10-02T17:16:13.488754787+05:30 [INFO] set status on client 0: Authenticating
2024-10-02T17:16:13.48911801+05:30 [INFO] perform;
2024-10-02T17:16:13.489230551+05:30 [INFO] authenticate;
2024-10-02T17:16:13.489356538+05:30 [INFO] read_command;
2024-10-02T17:16:13.489753484+05:30 [INFO] socket reader;
2024-10-02T17:16:13.49138443+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.491486622+05:30 [INFO] Opening main db
2024-10-02T17:16:13.492577953+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.493054389+05:30 [INFO] Opening main db
2024-10-02T17:16:13.494243203+05:30 [INFO] new;
2024-10-02T17:16:13.495109482+05:30 [INFO] keep_updated;
2024-10-02T17:16:13.512184497+05:30 [INFO] set status on client 235766: Connecting
2024-10-02T17:16:13.513154951+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.516915159+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.517252483+05:30 [ERROR] No worktree for path: "/usr/share/gnome-shell/extensions/pop-shell@system76.com/dark.css"
2024-10-02T17:16:13.536107725+05:30 [INFO] attempting to start language server "tailwindcss-language-server", path: "/home/tanay/Dev/Repos/FlowDoro", id: 1
2024-10-02T17:16:13.536274208+05:30 [INFO] Initializing default prettier with plugins {}
2024-10-02T17:16:13.536505222+05:30 [INFO] Initializing default prettier with plugins {}
2024-10-02T17:16:13.538876249+05:30 [INFO] Initializing default prettier with plugins {}
2024-10-02T17:16:13.539359938+05:30 [INFO] fetching latest version of language server "tailwindcss-language-server"
2024-10-02T17:16:13.53947767+05:30 [INFO] Node runtime install_if_needed
2024-10-02T17:16:13.543439897+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.543571144+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.582768352+05:30 [INFO] Using surface present mode MAILBOX
2024-10-02T17:16:13.585694743+05:30 [WARN] Unable to forbid exclusive full screen
2024-10-02T17:16:13.605076214+05:30 [INFO] set environment variables from shell:/bin/zsh, path:/home/tanay/bin:/home/tanay/.local/bin:/usr/local/bin:/home/tanay/bin:/home/tanay/.local/bin:/usr/local/bin:/home/tanay/.cargo/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin
2024-10-02T17:16:13.632024909+05:30 [INFO] attempting to start language server "vtsls", path: "/home/tanay/Dev/Repos/FlowDoro", id: 2
2024-10-02T17:16:13.632195679+05:30 [INFO] attempting to start language server "eslint", path: "/home/tanay/Dev/Repos/FlowDoro", id: 3
2024-10-02T17:16:13.632354528+05:30 [INFO] Initializing default prettier with plugins {}
2024-10-02T17:16:13.634824571+05:30 [INFO] fetching latest version of language server "vtsls"
2024-10-02T17:16:13.63498891+05:30 [INFO] fetching latest version of language server "eslint"
2024-10-02T17:16:13.635104467+05:30 [INFO] downloading language server "eslint"
2024-10-02T17:16:13.635415622+05:30 [INFO] Initializing default prettier with plugins {}
2024-10-02T17:16:13.636583798+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.677186528+05:30 [INFO] Node runtime install_if_needed
2024-10-02T17:16:13.711740779+05:30 [INFO] Initializing default prettier with plugins {}
2024-10-02T17:16:13.711863139+05:30 [INFO] attempting to start language server "vscode-html-language-server", path: "/home/tanay/Dev/Repos/FlowDoro", id: 4
2024-10-02T17:16:13.712655698+05:30 [WARN] request completed with error: failed to connect to the server
2024-10-02T17:16:13.776308325+05:30 [INFO] Node runtime install_if_needed
2024-10-02T17:16:13.831519534+05:30 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2024-10-02T17:16:13.865197708+05:30 [INFO] starting language server process. binary path: "/home/tanay/.local/share/zed/node/node-v22.5.1-linux-x64/bin/node", working directory: "/", args: ["/home/tanay/.local/share/zed/copilot/copilot-v0.5.0/dist/agent.js", "--stdio"]
2024-10-02T17:16:13.865448861+05:30 [INFO] Node runtime install_if_needed
2024-10-02T17:16:13.957995418+05:30 [INFO] Node runtime install_if_needed
2024-10-02T17:16:14.053920166+05:30 [INFO] Node runtime install_if_needed
2024-10-02T17:16:14.143742773+05:30 [INFO] using project environment for language server "eslint", id: 3
2024-10-02T17:16:14.143900259+05:30 [INFO] Node runtime install_if_needed
2024-10-02T17:16:14.144025635+05:30 [INFO] starting language server process. binary path: "/home/tanay/.local/share/zed/node/node-v22.5.1-linux-x64/bin/node", working directory: "/home/tanay/Dev/Repos/FlowDoro", args: ["--max-old-space-size=8192", "/home/tanay/.local/share/zed/languages/eslint/vscode-eslint-release/2.4.4/vscode-eslint/server/out/eslintServer.js", "--stdio"]
2024-10-02T17:16:14.220753739+05:30 [INFO] Language server with id 0 sent unhandled notification LogMessage:
{
  "level": 0,
  "message": "[DEBUG] [agent] [2024-10-02T11:46:14.216Z] Agent service starting",
  "metadataStr": "[DEBUG] [agent] [2024-10-02T11:46:14.216Z]",
  "extra": [
    "Agent service starting"
  ]
}

</pre></details>

**Comment by widersky** [comment]
Quite recently, Zed took up over 2GB of RAM in my case. Despite everything, it is still software in a very development stage - anyway, such situations should not occur when it advertises itself as a very light editor.

**Comment by tanaybhomia** [comment]
Yeah I agree, Will see if this gets fixed in the recent updates, I hope they are working on this.

**Comment by JosephTLyons** [comment]
Hey @tanaybhomia, we have a handful of other issues about high RAM usage; do any of these cover what you are experiencing here? I'd like to consolidate the issues, if so:

- https://github.com/zed-industries/zed/issues/7361
- https://github.com/zed-industries/zed/issues/7939
- https://github.com/zed-industries/zed/issues/10389
- https://github.com/zed-industries/zed/issues/13552
- https://github.com/zed-industries/zed/issues/14833
- https://github.com/zed-industries/zed/issues/15692

**Comment by tanaybhomia** [comment]
I think the last issue listed no #15692, this is the closest to me.

**Comment by edfloreshz** [comment]
Why is Zed spawning so many instances of rust-analyzer?

![image](https://github.com/user-attachments/assets/3eae51ea-7d71-4e2e-8186-8c1c0d9433dd)

**Comment by Reapimus** [comment]
> Why is Zed spawning so many instances of rust-analyzer?
> 
> ![image](https://private-user-images.githubusercontent.com/22224438/378995149-3eae51ea-7d71-4e2e-8186-8c1c0d9433dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NTE0NjEsIm5iZiI6MTcyOTY1MTE2MSwicGF0aCI6Ii8yMjIyNDQzOC8zNzg5OTUxNDktM2VhZTUxZWEtN2Q3MS00ZTJlLTgxODYtOGMxYzBkOTQzM2RkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIzVDAyMzkyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyMzU1MTU5MWEzOGIyM2U4YjE5NzhiN2I1YTlmYzUzYzNjNzA0NGFiMWFjNzQ3MjI1MmQ5ODIzNTAxMTZlY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mZhHBmRpRGmWGawpFDFjdUm6XEliRX4uzffOGjQzJug)

I've also noticed something similar whenever Zed is using a lot of RAM, it always has multiple instances of every active LSP server running. It seems like Zed is starting up new instances of the LSPs without shutting down old ones for some reason?

**Comment by dy-tea** [comment]
![image](https://github.com/user-attachments/assets/8826e864-02c9-4fee-bcf1-3d9a6c449619)
Zed will occasionally freeze and consume upwards of 9GB of ram, no idea what the cause is.

**Comment by tanaybhomia** [comment]
9GB ??? How large is the project that you are running ?

**Comment by dy-tea** [comment]
> 9GB ??? How large is the project that you are running ?
![image](https://github.com/user-attachments/assets/652b3f8d-c00d-4a60-b267-97237b26f64d)

**Comment by tanaybhomia** [comment]
Have you checked that if it is spawning Rust analyzers like the comment has ?

**Comment by emakarov** [comment]
not sure why, not too many files opened but zed started to consume up to several hundreds of gigabytes, and it's quite reproduceable

<img width="419" alt="Screenshot 2024-12-17 at 9 58 43 PM" src="https://github.com/user-attachments/assets/25229147-749d-4527-9f1b-133cd4847e4b" />

**Comment by DragonDev1906** [comment]
> Why is Zed spawning so many instances of rust-analyzer?

> Have you checked that if it is spawning Rust analyzers like https://github.com/zed-industries/zed/issues/18673#issuecomment-2430108163 comment says

This is something I wish to know, too. I believe it is related to how many crates a workspace has:
- I have 26 `Cargo.toml` files (1 is the workspace root, 2 are not part of the workspace, 23 are listed as workspace members)
- Zed has spawned 27 rust-analyzer processes
- The result is that Zed + child processes currently use 20 GB of memory.

In case it helps, here is my lsp configuration:
```json
"lsp": {
    "rust-analyzer": {
      "initialization_options": {
        "cargo": {
          "features": "all"
        },
        "procMacro": {
          "enable": true
        },
        "diagnostics": {
          "disabled": ["non_snake_case"]
        },
        "checkOnSave": true
      }
    }
  }
```

Potentially relevant: #20767

@edfloreshz how many rust-analyzer processes do you have and how many `Cargo.toml` files are in your folder?
@maan2003 Since you seem to have the same issue: Same question to you

**Comment by maan2003** [comment]
I think it was fixed by https://github.com/zed-industries/zed/pull/21787

cc @ConradIrwin

**Comment by edfloreshz** [comment]
I have a single `Cargo.toml` with 7 rust-analyzer processes open at the moment.

**Comment by CharlesChen0823** [comment]
I think is relate #21906

**Comment by Ayanrocks** [comment]
This issue is getting serious it seems. I open log files ranging from 50mb-200mb. And ZED consumes 19GB of memory for some reason on my 16gb MBP. Also if you keep it open it increases constantly. 

And another major issue is when I click to close the file. Entire Zed becomes unresponsive in Mac with that colorful circle kept on spinning. And I have to kill Zed to close it. 

I tried to do this in vscode and the ram usage wasn't too much. But This Ram issue is getting out of hand now. Every update I'm waiting for this issue to be listed in the changelog. 😞 .

Hope the team takes this as priority as well.

**Comment by Ayanrocks** [comment]
> This issue is getting serious it seems. I open log files ranging from 50mb-200mb. And ZED consumes 19GB of memory for some reason on my 16gb MBP. Also if you keep it open it increases constantly.
> 
> And another major issue is when I click to close the file. Entire Zed becomes unresponsive in Mac with that colorful circle kept on spinning. And I have to kill Zed to close it.
> 
> I tried to do this in vscode and the ram usage wasn't too much. But This Ram issue is getting out of hand now. Every update I'm waiting for this issue to be listed in the changelog. 😞 .
> 
> Hope the team takes this as priority as well.

Since A picture is worth a thousand words, Here are some screenshots for proof. Only One instance of Zed is running and I've closed the log files that I've opened. Still it's consuming 5.5GB of memory for some reason.

![Screenshot 2025-01-08 at 10 57 32 PM](https://github.com/user-attachments/assets/73eb9810-85a7-46f2-ad5e-49056bcb4611)
<img width="1395" alt="Screenshot 2025-01-08 at 10 57 29 PM" src="https://github.com/user-attachments/assets/8b090d07-bbbb-4a87-ab25-da9a0593a61f" />

**Comment by maan2003** [comment]
I found a repro for this

https://github.com/zed-industries/zed/issues/20767#issuecomment-2613574099

**Comment by playbahn** [comment]
> not sure why, not too many files opened but zed started to consume up to several hundreds of gigabytes, and it's quite reproduceable
> <img alt="Screenshot 2024-12-17 at 9 58 43 PM" width="419" src="https://private-user-images.githubusercontent.com/395963/396525845-25229147-749d-4527-9f1b-133cd4847e4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyODY2MzcsIm5iZiI6MTc0MTI4NjMzNywicGF0aCI6Ii8zOTU5NjMvMzk2NTI1ODQ1LTI1MjI5MTQ3LTc0OWQtNDUyNy05ZjFiLTEzM2NkNDg0N2U0Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwNlQxODM4NTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iOGY1YmU1YjEzN2EwNjI4YTcxMDg1ZTg1NGI5YWIzYWJjNTViNmM5ZmNkYTQ3ODQ0YTRjYmQ2YjQxMWJmYmYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fNkA_7XG-5vO2ES2rv3iYB9EjUKPYxMXqDz4IydwUC0">

What are you using that has so much RAM to give to processes?

**Comment by emakarov** [comment]
Its on Macbook Air. I think it is including swap

**Comment by mmstick** [comment]
I saw that Zed is using an older version of cosmic-text on Linux (`version = "0.11.2"`). It should be updated to the latest version (`0.13.2`), which contains a lot of memory usage optimizations. I also didn't see any malloc tunables being used, so I'd highly recommend setting `M_MMAP_THRESHOLD` on Linux to a reasonable value on application startup. libcosmic has [malloc functions declared here](https://github.com/pop-os/libcosmic/blob/master/src/malloc.rs) which get called early on application startup [like so](https://github.com/pop-os/libcosmic/blob/master/src/app/mod.rs#L99-L102). GNU libc malloc also does not automatically trim its brk arenas, so it can be necessary to malloc call `malloc::trim(0)` after an expensive operation has freed up a lot of memory.

**Comment by mcobzarenco** [comment]
As of today, `zed` seems to spawn rust-analyzer instances (in a loop maybe?) until it runs out of RAM (I have 64 GiB) and Ubuntu crashes.

This is for compiling a hello world Bevy project:

https://github.com/user-attachments/assets/ac94edc4-4128-40b4-9f75-81894b129706

**Comment by V-Legendre** [comment]
I am experiencing the same problem here. Zed's node server consumption increases to 2-3GB for a small project.

<img width="547" alt="Image" src="https://github.com/user-attachments/assets/63d215cc-0f01-4a93-ae1e-efe10772f5f8" />

**Environment**
- Macbook Pro M3, 16GB
- OS version: Sequoia 15.3.1

**Comment by a0405u** [comment]
24GB of RAM when Zed was on background for an hour.

**Comment by edfloreshz** [comment]
Just received the following notification:

![Image](https://github.com/user-attachments/assets/f37577d2-2e5d-4180-9440-0ec7748456ea)

**Comment by cschuhen** [comment]
I've decided to try Zed and noticed the same issue. Below you can see the memory use. I wasn't even coding in it. I was just looking at a couple of log files. One was 159487 bytes over 2165 lines. The other just 9 lines, 1044 bytes. Both files, just RAW ASCII text. No file extensions on the files. It had crashed earlier in the day (actually OOM killed it) and I'd just opened these two files and left them running for a couple of hours.

2813656 cschuhen  20   0   32.4g  13.5g 171952 S  82.4  21.6  48:55.48 zed-editor

**Comment by cschuhen** [comment]
So, I experimented a little and may have found something. Running with --foreground, (after trying valgrind to see it just detach), I noticed error messages being constantly printed about not finding root paths:

2025-06-03T10:51:22+10:00 ERROR [worktree] failed to canonicalize root path: No such file or directory (os error 2)

 A google search pointed me that this 'may' be related to GIT(#7214 was top result).
So, I ran git init in /tmp/. That didn't seem to help.. Memory kept creeping up.

HOWEVER... If I then save that file, which until that point, didn't exist on the FS yet, I see no more errors and my memory usage appears to stop creeping up. So.. there seems to me a memory leak somewhere dealing with new files, not saved at all. 

**In other words, I can reproduce this issue with:** 
> zed --foreground ` mktemp  -u` 
And make sure not to save the buffer.

**Comment by negentropi** [comment]
interting. for now my is 120mb+16mb(autofil(zed)).
looking forward zed team findout the prob on those cases.

**Comment by emakarov** [comment]
so far I managed to get zed working without growing memory by disabling all autocomplete and prediction features with such 2 variables in settings

```
  "show_edit_predictions": false,
  "show_completions_on_input": false,
```
I mostly use python and typescript in zed

Not sure if these 2 features are the root cause, though

**Comment by gwillen** [comment]
> So, I experimented a little and may have found something. Running with --foreground, (after trying valgrind to see it just detach), I noticed error messages being constantly printed about not finding root paths:
> 
> 2025-06-03T10:51:22+10:00 ERROR [worktree] failed to canonicalize root path: No such file or directory (os error 2)
> 
> A google search pointed me that this 'may' be related to GIT([#7214](https://github.com/zed-industries/zed/issues/7214) was top result). So, I ran git init in /tmp/. That didn't seem to help.. Memory kept creeping up.
> 
> HOWEVER... If I then save that file, which until that point, didn't exist on the FS yet, I see no more errors and my memory usage appears to stop creeping up. So.. there seems to me a memory leak somewhere dealing with new files, not saved at all.
> 
> **In other words, I can reproduce this issue with:**
> 
> > zed --foreground ` mktemp  -u`
> > And make sure not to save the buffer.

@cschuhen I believe this is caused by #34864, which I just opened, and is a bug specific to opening nonexistent files in /tmp/. (I think the overall problem with this issue #18673 is related to leaks in Zed, probably multiple different ones; but the _specific_ manifestation that you're seeing is caused by those leaks being repeatedly triggered by another bug.)

**Comment by cvanelteren** [comment]
Zed is claming 19gb on my machine, is there a fix for this: 

<img width="573" height="111" alt="Image" src="https://github.com/user-attachments/assets/074bd818-4dd5-459a-aa60-a2576de87b68" />

**Comment by cvanelteren** [comment]
The issue pops up when opening a project with lots of file. I was able to tame the RAM usage by putting `.nox` in `.gitignore`

**Comment by milanmlft** [comment]
Clocked over 62 GB today

<img width="480" height="426" alt="Image" src="https://github.com/user-attachments/assets/c7753d00-ab58-43a4-9b8d-50913b3cc380" />

Project with mostly Go and Terraform files + some Python:

```
❯ tokei -s files
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Go                    256        24283        20047         1140         3096
 HCL                   215        12686        10711          315         1660
 YAML                   27         3322         3217           59           46
 Python                 21         1588         1243           22          323
 Plain Text             18           27            0           27            0
 HTML                   15          875          832            0           43
 Dockerfile             13          368          239           31           98
 Makefile               11          648          521            2          125
 JSON                    9         8712         8712            0            0
 Shell                   9          275          196           35           44
 SVG                     6          308          306            1            1
 JavaScript              4          133          119            1           13
 BASH                    2           31           22            4            5
 CSS                     2          143          139            3            1
 TOML                    2           98           86            1           11
-------------------------------------------------------------------------------
 Markdown               38         1715            0         1107          608
 |- BASH                 4           13           13            0            0
 |- Shell                1            4            4            0            0
 |- YAML                 1           12           10            2            0
 (Total)                           1744           27         1109          608
===============================================================================
 Total                 648        55212        46390         2748         6074
===============================================================================
```

**Comment by nullart2** [comment]
ok this means I will have to wait until this memory leak is fixed

**Comment by silomalo** [comment]
Today am clocking upto 54GB.  Am I missing some settings or plugin?

<img width="434" height="370" alt="Image" src="https://github.com/user-attachments/assets/75ead7d5-f60d-4744-9607-6416fb0ba319" />

**Comment by JaeAeich** [comment]
As much as I love to use Zed, its making my system unusable to a point where the so called "bloated" IDEs feel much better.

<img width="1779" height="1102" alt="Image" src="https://github.com/user-attachments/assets/5b8a0b76-2ce4-4d9b-928d-04b0c80475fb" />

**Comment by wezm** [comment]
I have a six projects open in Zed at the moment. The Zed process itself has been up for ~8 days (started 2025-10-21T13:39:15+10:00) and it's using almost 10Gb of RAM. I'm not sure if this is considered acceptable for the number of projects/files open, but it does seem like a lot.

<img width="1051" height="578" alt="Image" src="https://github.com/user-attachments/assets/202b3c3a-a7de-4f9c-88e7-bbc31b575b3c" />

**Comment by BezPowell** [comment]
On my Laptop, Zed quickly freezes the system (within a few minutes) by consuming too much RAM when opening a small project (< 20 HTML files). Monitoring it via gnome system monitor, I see it start around 2GB, then quickly jump to 3.6 before the system becomes unresponsive.

My system specs:
- **Hardware Model:**                              ASUSTeK COMPUTER INC. N550JV
- **Memory:**                                      8.0 GiB
- **Processor:**                                   Intel® Core™ i7-4700HQ × 8
- **Graphics:**                                    Intel® HD Graphics 4600 (HSW GT2)
- **Graphics 1:**                                  NVE7
- **Disk Capacity:**                               240.1 GB
- **OS Name:**                                     Fedora Linux 42.20251030.0 (Silverblue)

It is an older laptop (around 11 years), but still handles fairly intensive tasks such as compiling Rust and light gaming without problems.

**Comment by mfouesneau** [comment]
Zed is unusable today, using 60GB of memory after opening a project.

Zed 0.214.5
Mac os Tahoe 
Mac M2

**Comment by Sachin-Bhat** [comment]
Second this.
```bash
zed --version
lsb_release -a
Zed preview 0.215.2 3b733feb12db25fc4347a5f04b04c55a0f738621
LSB Version:    n/a
Distributor ID: Artix
Description:    Artix Linux
Release:        rolling
Codename:       plasma-dinit-20250908
```

**Comment by insunaa** [comment]
Yeah, I keep seeing exceptionally high memory usage with zed in C++ projects, too. (Version 0.216.1) AI garbage obviously disabled.

<img width="998" height="274" alt="Image" src="https://github.com/user-attachments/assets/34f544ba-ee82-4d69-ac0b-e6eb05e0c564" />
clangd isn't innocent, but it's peanuts in comparison

<img width="745" height="283" alt="Image" src="https://github.com/user-attachments/assets/51219a55-83b9-48a9-8fd8-fcfe42c42433" />

I have 64GB of system RAM and zed regularly lets usage shoot up to 95% with several gigabytes in Swap, too

**Comment by noboomu** [comment]
I've had zed + rust-analyzer devour 90gb of ram for 3 not large rust projects. All extensions disabled. I've had to run claude/gpt cli sessions in separate terminal applications for a few months now because of zed consistently crashing. This is especially frustrating as I usually have all of the projects open at the same time in different workspaces, when one zed crashes they all crash.  

Is there an ability to launch Zed in separate processes? At least this way when we are forced to kill one, it does not kill all. 

At this point I am reluctantly opening up vscode for the first time in over a year because it does not have this issue. I love zed and want it to win, but if we can't run modern workflows on it, that leaves little option.

**Comment by alexbangert** [comment]
I’m seeing what looks like the same class of problem on Linux, but the main offender is not the Zed UI process itself, it’s the Kotlin language server started by Zed.

  Current environment:
  - Linux
  - Zed running with Kotlin and Java projects open

  Observed system state:
  - RAM: 31 GiB total
  - Swap (zram): 12 GiB total
  - At the time of investigation, zram usage had grown to about 10.2 GiB
  - MemAvailable was still around 9.7 GiB

  Top memory consumers:
  - Zed Kotlin LSP: ~6.7 GiB RSS
  - Gradle daemon: ~2.3 GiB RSS
  - Java JDT LS: ~390 MiB RSS


  The Kotlin process is this one:
  `/home/alex/.local/share/zed/extensions/work/kotlin/kotlin-lsp-262.2310.0/jre/bin/java ... com.jetbrains.ls.kotlinLsp.KotlinLspServerKt --stdio`

I also captured a heap dump and checked the live JVM:

  - Heap dump file size: 565 MiB
  - `jcmd GC.class_histogram` total: about 434 MB of Java objects
  - `jcmd GC.heap_info`: heap used about 424 MiB
  - Metaspace used about 134 MiB
  - JVM flags show `MaxHeapSize=8384413696` (about 8 GiB)

  `pmap` shows several very large anonymous mappings in the Kotlin LSP process, with the biggest RSS entries around:
  - 1.91 GiB
  - 1.47 GiB
  - 1.01 GiB
  - 470 MiB
  - 429 MiB

  So this does not look like a simple Java heap leak. It looks more like native/off-heap memory growth or some other non-heap allocation pattern inside the Kotlin LSP process
  launched by Zed.

So this does not look like a simple Java heap leak. It looks more like native/off-heap memory growth or some other non-heap allocation pattern inside the Kotlin LSP process launched by Zed.

This is severe enough to create real memory pressure during normal development workflows, especially when combined with Android/Gradle tooling.

Same project opened in vscode consumes much lesser memory.

**Comment by mitanshu-reckonsys** [comment]
I am facing similar issues. 

<img width="1920" height="1200" alt="Image" src="https://github.com/user-attachments/assets/dafe7ff4-8a95-45bc-bdd2-5b8b7e55f8fb" />

[Zed.log](https://github.com/user-attachments/files/27196139/Zed.log)

**Comment by StrathCole** [comment]
Same here. Tried to use zed for the past two days and like it in general. But it keeps growing in memory usage until it finally freezes the whole PC (Debian 13) until the oom killer gets rid of it.

**Comment by Error-Cezar** [comment]
Bump, near a GB on this one
Arch system with 8GB ram
<img width="477" height="184" alt="Image" src="https://github.com/user-attachments/assets/cf9efb21-c9e6-4b10-a4a2-837fb2152985" />

**Comment by StrathCole** [comment]
For me I am experiencing very high ram usage (until PC freezes) occasionally. Often it happens when subagents are spawned or when large file searches are done by the agent(s). Language server ram usage is not very high (maybe 1-3GB) while zed often uses 32G+ or worst case crashes the machine with a usage of 70G+
The git repo in that case had not much changes, I disabled git blame (for testing) with no improvement effect.
There has not yet been a reliable way of reproducing it, it seems to depend on what the agents do.
Memory analysis via `ps` or other methods did not reveal anything helpful until now.

**Comment by advenimus** [comment]
Also having this issue, particularly when working on a SSH host with zed 

<img width="400" height="41" alt="Image" src="https://github.com/user-attachments/assets/bf081f97-da3a-4f05-bf7d-1a218e60b476" />

**Comment by insalt-glitch** [comment]
I am running on Linux (Mint). Even if I kill my language servers Zed uses more and more RAM until it crashes my computer (upwards of 9GB). How can I approach finding the bug(s)? There must be some memory leak(s)?

**Comment by VirxEC** [comment]
I've experienced this now on both Linux and MacOS. After a while of using agents, my RAM is filling up, and without fail when I toggle the "Disable AI" option, RAM usage goes back down to normal and frees several GB. I then turn the AI features back on and RAM usage remains normal for a little bit.

**Comment by r00q** [comment]
The issue related to ACP thread size is tracked here: https://github.com/zed-industries/zed/issues/60134
But I don't think anything is being done...
